src/Future.re

```diff
@@ -75,3 +75,74 @@ let tapError = (future, f) => future |. tap(r => switch(r) {
   | Belt.Result.Error(v) => f(v) |. ignore
   | Ok(_) => ()
 });
+
+let flatMapArray = (xs: array('a), f: 'a => t('b), maxConcurrent) =>
```

Comment:
This feels like tying a few things together: batching the future creation, executing the batches one after another, and collecting the results.

I'd have expected something like:

```reason

/* this is the part where I prepare my data for the concurrent execution */
module type List = {
  /* take a list of 'a, and an int for how many to put in each bucket */
  let chunks_of: (int, list('a)) => list(list('a));
};

let futures = [1, 2, 3, 4] |. List.map(x => () => Future.value(x)); /* this could also be lazy futures */
let maxConcurrency = 2;
let chunks: list(list(unit => Future.t(int)) = List.chunks_of(maxConcurrency, futures);
/* [ [() => Future(1), () => Future(2)], [() => Future(3), () => Future(4)] ] */

/* this is the part where I build up the computation */
module type Future = {
  let all: (list(t('a)), list('a) => 'b); /* from https://github.com/RationalJS/future/issues/16 */
};

let sequence: list(list(unit => Future.t('a))) => Future.t(list(list('a)) = chunks => {
  let rec consumer = (cs, acc) => {
    switch (cs) {
    | [] => Future.value(acc);
    | [h, ...t] => {
        /* initialize all futures in this chunk: this would be more elegant using Lazy/Force */
       Future.all(List.map(x => x(), h))
       |. Future.flatMap(results => consumer(t, [results, ...acc])))
      }
    }
  };
  consumer(chunks, []);
};

/* this is the part where you consume it */
later_values.get( values => Js.log(values) );
```

Of course this is just a sketch but it's a bit more elegant, it (from what I can see) won't throw any exceptions anywhere, it can still be optimized with mutable cursors and good ol' `while` loops if the performance needs it.

Just a thought! Good work @dvisztempacct on pushing for concurrency idioms in the library 🙌

Reply:
This sort of design also gives the allocator a much greater workout. My original implementation (now called `flatMapArrayUnsafe`) only uses a single array of length `n` and (I believe) a single closure scope and two functions with access to that scope.

@ostera do you think you could provide a full implementation of your proposal so I could more closely study its performance characteristics?

Thanks :)

Reply:
> @dvisztempacct: each "chunk" must be processed serially
Oh, perhaps I misinterpreted the idea behind your `maxConcurrent`. I thought it to mean that Future's should be batched in groups of at most `maxConcurrent`.

Naturally if that wasn't the goal the implementation above can be changed by replacing the `Future.sequence` call with a `Scheduler`/`TaskRunner` that can take care of this concern.

Thanks for pointing this out! 💯 

> @dvisztempacct: This sort of design also gives the allocator a much greater workout
If by allocator you mean whoever uses this library, yes. If you mean the runtime, you'd still be creating a function for every `xs[i] |. f |. get( x => ... )` 🤔. Optimizations on both ends can be made to reuse the functions (in your case you have everything already in the upper scope, so it should be trivial to extract). It'd be interesting to compare the compiled Javascript of both solutions.

In the end your implementation will run faster than mine by the nature of it's design. I'll extract a Scheduler from it and perhaps then we can compare more fairly 👍

Reply:
Thanks @ostera

Reply:
@ostera https://github.com/jonlaing/rationale/blob/master/src/Monad.re. Is this similar to what you are looking for?

Reply:
@scull7 essentially this bit -> https://github.com/jonlaing/rationale/blob/master/src/Monad.re#L48-L54

Reply:
But it's different from what @dvisztempacct implemented tho! His Future pool keeps a constant number of promises running, whereas using `all` or `all_ignore` will run them in sequence, one by one AFAICT.